### ISSUE TITLE
Review / Test Controller Framework for Early Stopping

### ISSUE THREAD

**Issue by alex-jw-brooks** [foundation-model-stack/fms-hf-tuning#247]
## Is your feature request related to a problem? Please describe.

Issue for reviewing recent controller framework changes / documentation and adding some short examples for testing the framework with early stopping.

Also captures any small changes needed as a result.

**Comment by alex-jw-brooks** [comment]
Some examples for testing the controller framework through the accelerate launch script in a release image that might be helpful are below.


First, create a directory `tc_test_resources` with the following files:
 
`tc_test_resources/twitter_complaints_small.json` - we'll use this as our training and/or validation data for validating behaviors.
```json
{"Tweet text":"@HMRCcustomers No this is my first job","ID":0,"Label":2,"text_label":"no complaint","output":"### Text: @HMRCcustomers No this is my first job\n\n### Label: no complaint"}
{"Tweet text":"@KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.","ID":1,"Label":2,"text_label":"no complaint","output":"### Text: @KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.\n\n### Label: no complaint"}
{"Tweet text":"If I can't get my 3rd pair of @beatsbydre powerbeats to work today I'm doneski man. This is a slap in my balls. Your next @Bose @BoseService","ID":2,"Label":1,"text_label":"complaint","output":"### Text: If I can't get my 3rd pair of @beatsbydre powerbeats to work today I'm doneski man. This is a slap in my balls. Your next @Bose @BoseService\n\n### Label: complaint"}
{"Tweet text":"@EE On Rosneath Arial having good upload and download speeds but terrible latency 200ms. Why is this.","ID":3,"Label":1,"text_label":"complaint","output":"### Text: @EE On Rosneath Arial having good upload and download speeds but terrible latency 200ms. Why is this.\n\n### Label: complaint"}
{"Tweet text":"Couples wallpaper, so cute. :) #BrothersAtHome","ID":4,"Label":2,"text_label":"no complaint","output":"### Text: Couples wallpaper, so cute. :) #BrothersAtHome\n\n### Label: no complaint"}
{"Tweet text":"@mckelldogs This might just be me, but-- eyedrops? Artificial tears are so useful when you're sleep-deprived and sp\u2026 https:\/\/t.co\/WRtNsokblG","ID":5,"Label":2,"text_label":"no complaint","output":"### Text: @mckelldogs This might just be me, but-- eyedrops? Artificial tears are so useful when you're sleep-deprived and sp\u2026 https:\/\/t.co\/WRtNsokblG\n\n### Label: no complaint"}
{"Tweet text":"@Yelp can we get the exact calculations for a business rating (for example if its 4 stars but actually 4.2) or do we use a 3rd party site?","ID":6,"Label":2,"text_label":"no complaint","output":"### Text: @Yelp can we get the exact calculations for a business rating (for example if its 4 stars but actually 4.2) or do we use a 3rd party site?\n\n### Label: no complaint"}
{"Tweet text":"@nationalgridus I have no water and the bill is current and paid. Can you do something about this?","ID":7,"Label":1,"text_label":"complaint","output":"### Text: @nationalgridus I have no water and the bill is current and paid. Can you do something about this?\n\n### Label: complaint"}
{"Tweet text":"Never shopping at @MACcosmetics again. Every time I go in there, their employees are super rude\/condescending. I'll take my $$ to @Sephora","ID":8,"Label":1,"text_label":"complaint","output":"### Text: Never shopping at @MACcosmetics again. Every time I go in there, their employees are super rude\/condescending. I'll take my $$ to @Sephora\n\n### Label: complaint"}
{"Tweet text":"@JenniferTilly Merry Christmas to as well. You get more stunning every year \ufffd\ufffd","ID":9,"Label":2,"text_label":"no complaint","output":"### Text: @JenniferTilly Merry Christmas to as well. You get more stunning every year \ufffd\ufffd\n\n### Label: no complaint"}
```


`tc_test_resources/exit_if_epoch_num_exceeds_one.yaml` - this config will break out of training after the second epoch ends. If we pass a high number of training epochs, it should still only run 2.
```yaml
controller_metrics:
  - name: trainer_state
    class: TrainingState
  - name: evalmetric
    class: EvalMetrics
controllers:
  - name: exit_if_epoch_num_exceeds_one
    triggers:
      - on_epoch_end
    rule: trainer_state["epoch"] > 1
    operations:
      - hfcontrols.should_training_stop
```


`tc_test_resources/exit_based_on_historical_loss.yaml` - this config will break out of training if the evaluation loss is greater than 2 for more than two epochs after the metric window has started (which is a bit silly, but it's just to test the exit functionality). This should exit after the 4th epoch, because of the timing of when the window has something in it + the time to exceed the patience condition for the loss check.
```yaml
controller_metrics:
  - name: eval_loss_window
    class: HistoryBasedMetric
    arguments:
      window_size: 2
controllers:
  - name: epoch_level_eval_loss_patience
    triggers:
      - on_epoch_end
    rule: len(eval_loss_window["metrics"]) > 0 and eval_loss_window["metrics"]["eval_loss"][-1] > 2
    patience:
      patience_threshold: 2
    operations:
      - hfcontrols.should_training_stop
```

`tc_test_resources/config.json`  - Config to be used for testing with a tiny model. The main things to note here are that we specify 20 train epochs and should break before then if stopping early, and that we specify the path to the mounted yaml config for the trainer controller here. 
```
{
    "model_name_or_path": "Maykeye/TinyLLama-v0",
    "tokenizer_name_or_path": "Maykeye/TinyLLama-v0",
    "training_data_path": "/tc_test_resources/twitter_complaints_small.json",
    "validation_data_path": "/tc_test_resources/twitter_complaints_small.json",
    "output_dir": "/tc_test_resources/tc_test_model",
    "num_train_epochs": 20.0,
    "per_device_train_batch_size": 4,
    "per_device_eval_batch_size": 4,
    "gradient_accumulation_steps": 1,
    "evaluation_strategy": "epoch",
    "save_strategy": "epoch",
    "learning_rate": 1e-5,
    "response_template": "\n### Label:",
    "dataset_text_field": "output",
    "peft_method": "pt",
    "num_virtal_tokens": 100,
    "prompt_tuning_init_text": "Classify if the tweet is a complaint or not:",
    "trainer_controller_config_file": "/tc_test_resources/exit_if_epoch_num_exceeds_one.yaml",
    "use_flash_attn": false
}
```

Next, start an interactive shell in your image with this directory mounted; for example, if you built a dev wheel, you'd do something like this:
```bash
TC_RESOURCE_DIR={absolute path to your tc_test_resources dir}
docker run -it -v $TC_RESOURCE_DIR:/tc_test_resources docker.io/library/fms-hf-tuning:dev bash
```

Now you should be able to run the launch script normally, e.g.,

```bash
export SFT_TRAINER_CONFIG_JSON_PATH=/tc_test_resources/config.json
python accelerate_launch.py 
```

### Validating Correctness
For `"trainer_controller_config_file": "/tc_test_resources/exit_if_epoch_num_exceeds_one.yaml"`, we expect to terminate after one epoch.

**Comment by Ssukriti** [comment]
Thanks for the excellent record @alex-jw-brooks . I will close the issue and we will pass it as reference to concerned teams as needed
